### ISSUE TITLE
crash on macos

### ISSUE THREAD

**Issue by Silvercast** [Yubico/libfido2#26]
The following snipet worked on `99a9be59b77fc9e6513a2953c222b44b0dbbe91e` but not on latest master `446408aab849d60aeec92558eeabac9aea96636f`

```
fido_assert_set_clientdata_hash(assert, cdh, sizeof(cdh));
fido_assert_set_rp(assert, rpId);
fido_assert_set_extensions(assert, 0);
fido_assert_set_options(assert, true, false);
fido_assert_allow_cred(assert, credid->buffer, credid->length);
fido_dev_get_assert(device, assert, NULL);
```

It just crashes real hard :(

**Comment by martelletto** [comment]
Thanks, I will have a look. Would you be able to capture a stack trace?

**Comment by Silvercast** [comment]
sadly no, but I managed to reproduce very time. It only occurs on old U2F devices like YubiKey NEO Firmware 3.5. On new FIDO2 devices it doesn't crash.

**Comment by martelletto** [comment]
Thanks, that already narrows down the problem. 

Unfortunately, I could not reproduce it on a YubiKey NEO 3.5.0 using examples/assert and tools/fido2-assert, with libfido2 and its dependencies built with ASAN. I found what seems to be a memory leak in authdata_fake() (in u2f.c), but that seems unrelated. It certainly does not explain what you are seeing. I will give it a closer look next week, as I am currently on vacation.

In the meantime, I will need as much information about your runtime as possible. Could you tell me the values of sizeof(cdh) and credid->length, and the length of rpId? Would it be possible to run your application with the environment variable FIDO_DEBUG set to 1?

**Comment by Silvercast** [comment]
thanks, I will try to compile and run `examples/cred_assert`. could it be because we use OpenSSL FIPS ?

setting FIDO_DEBUG to 1 gives some new info

```
ConsumerAcis_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
tx: d=0x600000962e90, cmd=0x86, buf=0x600000962e90, count=8
  f6 3a 48 32 5b 62 bc c0 
rx: initiation frame at 0x70000fd28390, len 64
  ff ff ff ff 86 00 11 f6 3a 48 32 5b 62 bc c0 04 
  00 82 01 02 01 01 00 01 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
tx: d=0x600000962e90, cmd=0x83, buf=0x600001c092d2, count=136
  00 02 07 00 00 00 81 ff ff ff ff ff ff ff ff ff 
  ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
  ff ff ff ff ff ff ff a7 ce 2b ba c0 34 8f 01 34 
  0f 9e 5d d3 e7 93 28 61 2b 05 67 49 0c d5 dc 88 
  45 bf a6 5a 29 4b ad 40 cb c4 46 49 01 70 d5 cc 
  12 26 09 31 46 9e b9 63 e3 7d 3c b6 ac 8a 88 60 
  dc dc 4a 79 e9 50 fe a9 e3 06 bb 12 9e bb 24 5d 
  52 61 27 3a 0b c2 5e 00 42 ee d9 1d 48 3e 08 4d 
  1d 79 c3 f5 43 c1 c1 etion]
```

**Comment by Silvercast** [comment]
size of cdh : 32 (SHA256_DIGEST_LENGTH)
rpid : appgate
credid->length : 64

**Comment by Silvercast** [comment]
I'm on macOS latest and we try to statically link libfido2 to a Xamarin project. I saw some complains in Console. Maybe an issue in hid_osx ?

**Comment by martelletto** [comment]
Thanks. I doubt the problem is related to OpenSSL. From the output of FIDO_DEBUG=1, it seems the application is crashing either during, or shortly after the tx op in key_lookup() [1]. This tx operation is followed by a rx, which is nowhere in the debug output. The length of 136 is consistent with what I would expect on a x86_64 platform. However, only 135 payload bytes are printed, which is suspicious.

Yes, this definitely looks like a problem in hid_osx.c. However, I would like to first understand where the 136th byte went. Could you double-check that indeed only 135 bytes are printed? Perhaps the output was mangled.

[1] https://github.com/Yubico/libfido2/blob/8a98814249ba984725cb04016837be7b9fbdfced/src/u2f.c#L191

**Comment by Silvercast** [comment]
I left my mac at the other place so I can't provide the log on macOS today. But here is some other good news. The crash only appears on
1. U2F keys for example YubiKey NEO Firmware 3.5
2. macOS
3. New version of libfido2

It doesn't crash on Linux nor Windows. so definitely macOS hid_osx.

**Comment by martelletto** [comment]
Thanks for the investigation. I also don't have access to a MacOS machine
at the moment, but will on Monday.

**Comment by martelletto** [comment]
Hi,

Can you check whether the diff below fixes the problem for you?

-p.

```diff
diff --git src/hid_osx.c src/hid_osx.c
index 40c502b..3b158f7 100644
--- src/hid_osx.c
+++ src/hid_osx.c
@@ -368,6 +368,8 @@ hid_read(void *handle, unsigned char *buf, size_t len, int ms)
 
        IOHIDDeviceUnscheduleFromRunLoop(dev->ref, CFRunLoopGetCurrent(),
            dev->loop_id);
+       IOHIDDeviceRegisterInputReportCallback(dev->ref, buf, len, NULL,
+           &status);
 
        if (status == -1) {
                log_debug("%s: io error", __func__);
```

**Comment by Silvercast** [comment]
Thanks, your patch above plus the commit `ebc9e2b11729deaef6c66a20fb2f1aa33eb024ae` partially fixes the crash on `fido_dev_get_assert`. 

- It doesn't seem to crash on cold start (turn off the program, turn on then call `fido_dev_get_assert` directly) but it crashes when I `fido_dev_make_cred` then `fido_dev_get_assert`.
- Without commit `ebc9e2b11729deaef6c66a20fb2f1aa33eb024ae` it crashes on cold start.

So there is memory leak + buffer overflow somewhere on macOS.

**Comment by martelletto** [comment]
Ok, so you are still seeing a crash when you do a fido_dev_make_cred() followed by fido_dev_get_assert(). It is probably exploding in the same part of the code. I will try to reproduce it locally and give it a closer look.

**Comment by Silvercast** [comment]
I wasn't giving fully correct info, it crashes randomly still.

**Comment by Silvercast** [comment]
The following changes seem to fix the crash for me completely. However I'm not confident about removing the check of status.

```
diff --git a/src/hid_osx.c b/src/hid_osx.c
index 40c502b..3b516ec 100644
--- a/src/hid_osx.c
+++ b/src/hid_osx.c
@@ -333,14 +333,13 @@ read_callback(void *context, IOReturn result, void *dev, IOHIDReportType type,

        (void)dev;
        (void)report;
+       (void)status;

        if (result != kIOReturnSuccess || type != kIOHIDReportTypeInput ||
            report_id != 0 || report_len != REPORT_LEN - 1) {
                log_debug("%s: io error", __func__);
                return;
        }
-
-       *status = 0;
 }

 int
@@ -348,7 +347,6 @@ hid_read(void *handle, unsigned char *buf, size_t len, int ms)
 {
        struct dev              *dev = handle;
        CFRunLoopRunResult      r;
-       int                     status = -1;

        (void)ms; /* XXX */

@@ -358,22 +356,17 @@ hid_read(void *handle, unsigned char *buf, size_t len, int ms)
        }

        IOHIDDeviceRegisterInputReportCallback(dev->ref, buf, len,
-           &read_callback, &status);
+           &read_callback, NULL);
         IOHIDDeviceScheduleWithRunLoop(dev->ref, CFRunLoopGetCurrent(),
            dev->loop_id);

        do
                r = CFRunLoopRunInMode(dev->loop_id, 0.003, true);
-       while (status == -1 && r != kCFRunLoopRunHandledSource);
+       while (r != kCFRunLoopRunHandledSource);

        IOHIDDeviceUnscheduleFromRunLoop(dev->ref, CFRunLoopGetCurrent(),
            dev->loop_id);

-       if (status == -1) {
-               log_debug("%s: io error", __func__);
-               return (-1);
-       }
-
        return (REPORT_LEN - 1);
 }
```

**Comment by martelletto** [comment]
Fixed in 06abce98.
